Update Working with Dependencies -> Declaring Dependencies Section (#30507)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/dep-man/declaring-depedencies/declaring_configurations.adoc

// Copyright (C) 2023 Gradle, Inc.
//
// Licensed under the Creative Commons Attribution-Noncommercial-ShareAlike 4.0 International License.;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://creativecommons.org/licenses/by-nc-sa/4.0/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[dependency-configurations]]
= Declaring Dependency Configurations

[[sec:what-are-dependency-configurations]]
In Gradle, dependencies are associated with specific scopes, such as compile-time or runtime.
These scopes are represented by configurations, each identified by a unique name.

image::dependency-management-configurations.png[]

> Gradle plugins often add pre-defined link:{groovyDslPath}/org.gradle.api.artifacts.Configuration.html[configuration] to your project.

Reviewer: ```suggestion
Gradle plugins often add pre-defined link:{groovyDslPath}/org.gradle.api.artifacts.Configuration.html[configurations] to your project.
```

Author: Done.

====
include::sample[dir="snippets/dependencyManagement/customizingResolution-classifierSubstitutionRule/kotlin",files="consumer/build.gradle.kts[tags=dependencies]"]
include::sample[dir="snippets/dependencyManagement/customizingResolution-classifierSubstitutionRule/groovy",files="consumer/build.gradle[tags=dependencies]"]
====

This example highlights project and module dependencies declared on the `implementation` configuration.
In a Java project, the `implementation` configuration is used to specify dependencies required to compile and run the main source code.
See the <<java_plugin.adoc#sec:java_plugin_and_dependency_management,Java plugin documentation>> for details.

[[sub:config-inheritance-composition]]
== Understanding configuration inheritance

Configurations can inherit from other configurations, creating an inheritance hierarchy.
Child configurations inherit all dependencies declared in their parent configurations.

Configuration inheritance is commonly used by Gradle core plugins, such as the <<java_plugin.adoc#sec:java_plugin_and_dependency_management,Java plugin>>.

> For instance, the `testImplementation` configuration extends the `implementation` configuration.

Reviewer: This is actually a pattern we'd like to move away from. 

Instead, we can say the `implementation` configuration extends from the `api` configuration

Reviewer: We should also update the below text that talks about JUnit

Author: TODO: This needs a small discussion on what good examples we have.

Author: Complete overhaul of the page complete following discussion.

image::dependency-management-configuration-inheritance.png[]

This hierarchy ensures that dependencies needed for both compiling tests and the source code under test are included.
For example, a Java project using JUnit for tests might also need Guava if it's used in the production code.

> Internally, configurations form an inheritance hierarchy using the link:{groovyDslPath}/org.gradle.api.artifacts.Configuration.html#org.gradle.api.artifacts.Configuration:extendsFrom(org.gradle.api.artifacts.Configuration++[]++)[`Configuration.extendsFrom(Configuration...)`] method.

Reviewer: Are those `++` in the URL intentional?

Author: Not intended. Not sure what happened to the URL. Fixed.

> Internally, configurations form an inheritance hierarchy using the link:{groovyDslPath}/org.gradle.api.artifacts.Configuration.html#org.gradle.api.artifacts.Configuration:extendsFrom(org.gradle.api.artifacts.Configuration++[]++)[`Configuration.extendsFrom(Configuration...)`] method.

Reviewer: Why include the `Internally`? To me that makes it seem this is a method that is only used internally and not something that a user would call

Author: Fixed.

A configuration can extend any other configuration, regardless of how it is defined in the build script or plugin.

> For instance, if you want to create a suite of smoke tests that verify a web service endpoint using JUnit, you can define a new configuration named `smokeTest` that extends from `testImplementation` to reuse the existing test framework dependency:

Reviewer: With test suites we explicitly avoided this sort of pattern so I'm not sure its a great way to showcase extendsFrom in the docs

Author: TODO: This needs a small discussion on what good examples we have.

Author: Complete overhaul of the page complete following discussion.

====
include::sample[dir="snippets/dependencyManagement/definingUsingConfigurations-inheritance/kotlin",files="build.gradle.kts[tags=configuration-definition]"]
include::sample[dir="snippets/dependencyManagement/definingUsingConfigurations-inheritance/groovy",files="build.gradle[tags=configuration-definition]"]
====

NOTE: Configurations can only extend configurations within the same project.

[[sec:defining-custom-configurations]]
== Creating custom configurations

> You can define custom configurations to separate the scope of dependencies for specific purposes.

Reviewer: ```suggestion
You can define custom configurations to declare separate scopes of dependencies for specific purposes.
```

Author: Done.

> For example, to use the link:https://tomcat.apache.org/tomcat-9.0-doc/jasper-howto.html[Jasper Ant task] for pre-compiling JSP files without including it in the classpath for compiling your source code, you can create and use a custom configuration:

Reviewer: I would definitely not use this sample here. This showcases more how to integrate with ANT than anything. I really think we should use another example here.

Author: Updated to a better example.

====
include::sample[dir="snippets/dependencyManagement/definingUsingConfigurations-custom/kotlin",files="build.gradle.kts[tags=custom-configuration]"]
include::sample[dir="snippets/dependencyManagement/definingUsingConfigurations-custom/groovy",files="build.gradle[tags=custom-configuration]"]
====

> You can manage custom configurations using the `configurations` object.

Reviewer: ```suggestion
You can manage custom configurations using the `configurations` block.
```

Author: Done.

Configurations can have names and can extend each other.
For more details, refer to the link:{groovyDslPath}/org.gradle.api.artifacts.ConfigurationContainer.html[`ConfigurationContainer`] API.

> Remember that configurations are intended to be used for a single role: declaring dependencies, performing resolution, or defining consumable variants.

Reviewer: You say "remember" here, but has this information been described anywhere before? 

I think we should link to that "prior knowledge" if we're going to use language like this

Author: This is a leftover from the old docs and has now been rewritten.

====

*Configurations used for declaring dependencies* define and organize the dependencies required for different aspects of a project.

=== Configurations for consumers (resolvable configuration)

To specify how dependencies are used, you can create companion configurations:

====
include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=concrete-classpath]"]
include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",files="build.gradle[tags=concrete-classpath]"]
====

Now, `someConfiguration` declares dependencies, while `compileClasspath` and `runtimeClasspath` are resolvable configurations for specific purposes.
When resolved they should contain the compile classpath, and the runtime classpath of the application respectively.

*Resolvable configurations* are those that can be resolved to produce a set of files or artifacts.
These configurations are used to define the classpath for different stages of a build process, such as compilation or runtime.

=== Configurations for producers (consumable configuration)

To specify what is exposed to consumers, you can create consumable configurations:

====

> include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=setup-configurations]"]

Reviewer: In this example, there are some assertions in the code. I'm not exactly sure why they are there, but since users are not going to write those assertions themselves, we should not include it in the code sample.

Author: The assertions were removed, and they were not used by any test.

include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",files="build.gradle[tags=setup-configurations]"]
====

Here, a library exposes configurations for its API and runtime.
The `apiElements` configuration provides the API for compilation, while the `runtimeElements` configuration offers the runtime dependencies.
These configurations are consumable but not resolvable, as indicated by the `canBeConsumed` and `canBeResolved` flags.

*Consumable Configurations* expose artifacts for other projects to consume.
They are not intended to be resolved directly but are meant to provide dependencies or resources that other projects can depend on.

=== Configuration flags and roles

Configurations have two key flags:

> - **`canBeResolved`:** Determines if a configuration can be resolved to a dependency graph. A configuration with `canBeResolved` set to `true` can be resolved, while one with `false` is only for declaring dependencies.

Reviewer: ```suggestion
- **`canBeResolved`:** Indicates that this configuration is intended for resolving a set of dependencies into a dependency graph.
```

I don't think the second part is necessarily true. `false` for canBeResolved can also indicate that this is intended for consumption

Author: Done.

> - **`canBeConsumed`:** Indicates if a configuration is meant for consumers. A configuration with `canBeConsumed` set to `true` is exposed to other projects.

Reviewer: ```suggestion
- **`canBeConsumed`:** Indicates that this configuration is intended for exposing artifacts outside this project.
```

Author: Done.

> 

Reviewer: We also have an incubating `isCanBeDeclared` which indicates whether you can declare dependencies on the configuration: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/Configuration.html#isCanBeDeclared()

In general, configurations should only have one of these flags enabled. 
A declarable configuration should not be resolvable or consumable.
A consumable configuration should not be declarable or resolvable.
A resolvable configuration should not be declarable or consumable.

We should probably add a section describing the incubating configuration factory methods on https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ConfigurationContainer.html (resolvable(), consumable(), dependencyScope()), which create configurations in these roles automatically, so users don't need to set the flags.

Author: Added but should be checked for accuracy.

> In short, a configuration's role is determined by the `canBeResolved` and `canBeConsumed` flag combinations:

Reviewer: This section should be updated to talk about all three flags.

Author: Done.

- `exposedApi` + `exposedRuntime`: Consumable configurations that expose artifacts (API and runtime) to other projects, but aren’t meant for internal resolution.

This setup mimics the behavior of the `implementation`, `compileClasspath`, `runtimeClasspath`, `apiElements`, and `runtimeElements` configurations in the <<java_plugin.adoc#java_plugin,Java plugin>>.

[[sec:defining-custom-configurations]]
== Creating custom configurations

You can define custom configurations to declare separate scopes of dependencies for specific purposes.

Suppose you want to generate Javadocs with AsciiDoc formatting embedded within your Java source code comments.
By setting up the `asciidoclet` configuration, you enable Gradle to use Asciidoclet, allowing your Javadoc task to produce HTML documentation with enhanced formatting options:

====
include::sample[dir="snippets/java/basic/kotlin",files="build.gradle.kts[tags=create-custom-doclet]"]
include::sample[dir="snippets/java/basic/groovy",files="build.gradle[tags=create-custom-doclet]"]
====

You can manage custom configurations using the `configurations` block.

> Configurations can have names and can extend each other.

Reviewer 2: ❗ "Can" have names isn't correct.  They "must" have names, even detached configurations have autogenerated names.

Author: Fixed.

For more details, refer to the link:{groovyDslPath}/org.gradle.api.artifacts.ConfigurationContainer.html[`ConfigurationContainer`] API.

NOTE: Configurations are intended to be used for a single role: declaring dependencies, performing resolution, or defining consumable variants.

There are three main use cases for creating custom configurations:

1. **API/Implementation Separation:** Create custom configurations to separate API dependencies (exposed to consumers) from implementation dependencies (used internally during compilation or runtime).
- You might create an `api` configuration for libraries that consumers will depend on, and an `implementation` configuration for libraries that are only needed internally. The `api` configuration is typically consumed by downstream projects, while `implementation` dependencies are hidden from consumers but used internally.

> - This separation ensures that your project maintains clean boundaries between its API and internal dependencies.

Reviewer 2: ```suggestion
- This separation ensures that your project maintains clean boundaries between its public API and strictly internal mechanisms.
```

Author: Fixed.

> 2. **Resolvable Configuration from Dependency Configuration**: Create a custom resolvable configuration to resolve specific sets of dependencies, like classpaths, at various build stages.

Reviewer 2: 🤔 I don't understand the "from Dependendency Configuration" title here.  What is this trying to say?  Everything after that makes sense.

- This allows fine-grained control over which dependencies are available during different build phases, such as compilation or testing.

3. **Consumable Configuration from Dependency Configuration**: Create a custom consumable configuration to expose artifacts or dependencies for other projects to consume, typically when your project produces artifacts like JARs.
- You might create an `exposedApi` configuration to expose the API dependencies of your project for consumption by other projects. Similarly, a `runtimeElements` configuration could be created to expose the runtime dependencies or artifacts that other projects need.
- Consumable configurations ensure that only the necessary artifacts or dependencies are exposed to consumers.

=== Configuration API incubating methods

Several incubating factory methods—`resolvable()`, `consumable()`, and `dependencyScope()`—within the link:{javadocPath}/org/gradle/api/artifacts/ConfigurationContainer.html[`ConfigurationContainer`] API can be used to simplify the creation of configurations with specific roles.

> These methods help users avoid manually setting various configuration flags, streamlining the process and ensuring consistency:

Reviewer 2: ```suggestion
These methods help build authors document the purpose of a configuration and avoid manually setting various configuration flags, streamlining the process and ensuring consistency:
```

Author: Fixed.

- **`resolvable()`**: Creates a configuration intended for resolving dependencies. This means the configuration can be used to resolve dependencies but not consumed by other projects.
- **`consumable()`**: Creates a configuration meant to be consumed by other projects but not used to resolve dependencies itself.
- **`dependencyScope()`**: Creates a configuration that establishes a dependency scope, setting up the necessary properties to act both as a consumer and provider, depending on the use case.

[[sub:config-inheritance-composition]]
== Configuration inheritance

Configurations can inherit from other configurations, creating an inheritance hierarchy.

Configurations form an inheritance hierarchy using the link:{groovyDslPath}/org.gradle.api.artifacts.Configuration.html#org.gradle.api.artifacts.Configuration:extendsFrom)[`Configuration.extendsFrom(Configuration...)`] method.

> A configuration can extend any other configuration, regardless of how it is defined in the build script or plugin.

Author: Added.

> NOTE: Configurations can only extend configurations within the same project.

Author: Added tip.

> When extending a configuration, the new configuration inherits:

Author: Done.
